### PR TITLE
Parallel spending counters

### PR DESCRIPTION
**Pull request by enzoc4**
first draft, still needs a few adjustments.

and also the java and plugin updates.

### CODE REVIEW

File: wallet/src/account.rs [hunk 1]
```diff
@@ -20,7 +22,18 @@ pub struct Wallet {
 #[derive(Debug)]
 pub struct State {
     value: Value,
-    counter: SpendingCounter,
+    counters: Vec<SpendingCounter>,
```
Comment (Zeegomo) [inline wallet/src/account.rs:25]: Is there a reason you're not using `SpendingCounterIncreasing`?

Comment (enzoc4) [reply wallet/src/account.rs:25]: I was using it initially, but I was a bit unsure of a few minor things and replaced it with this, as the logic is really simple anyway. Mostly it seemed a bit coupled with the ledger.

For example, I was considering on relaxing the property of the constructor taking the spending counters sorted, as they have the lane inside anyway (I didn't do it, but I think could make sense). I still find strange that the argument is called `set` but still requires the outputs to be sorted. Also, `get_valid_counters` doesn't make any promises about the order.

Also, `next_verify` being mutable and returning a `LedgerError` is a bit strange. I know I can just ignore the error type and replace it with the error type in here, but... I don't know. And I guess cloning and then mutating is fine, but it seems more natural to just build the new vec.

I don't really have a strong posture on this though, I just wasn't sure about creating unnecessary coupling, if you think it's better to use that then it's fine for me.

Comment (Zeegomo) [reply wallet/src/account.rs:25]: I just have basically the same problem in testing wallets. On one side it's nice to have things tied to the ledger (e.g, you don't accidentally use more lanes that are available), but as you noted there are a few downsides. 
However, IIRC also `sub` operations on wallet value return a `LedgerError`

File: bindings/wallet-cordova/src/ios/WalletPlugin.m [hunk 2]
```diff
@@ -237,13 +246,25 @@ - (void)WALLET_SET_STATE:(CDVInvokedUrlCommand*)command
 
     NSString* wallet_ptr_raw = [command.arguments objectAtIndex:0];
     NSString* value_raw = [command.arguments objectAtIndex:1];
-    NSString* counter_raw = [command.arguments objectAtIndex:2];
+    NSArray* counters_raw = [command.arguments objectAtIndex:2];
```
Comment (enzoc4) [inline bindings/wallet-cordova/src/ios/WalletPlugin.m:249]: I don't know if this works tbh, as there is no other place in the code where we take a json array

File: wallet/src/account.rs (hunk 1 above)
Comment (TenStrings) [reply wallet/src/account.rs]: I decided to use it anyway to keep things consistent, the issues that I had were really minor anyway. If we need more flexibility in the future we can just change it.

Comment (enzoc4) [reply wallet/src/account.rs]: Decided to use it anyway, the issues that I had were really minor (and I actually had misunderstood part of the api), we can just change it in the future if we need more flexibility or something.

File: bindings/wallet-cordova/src/ios/WalletPlugin.m (hunk 2 above)
Comment (eugene-babichenko) [reply bindings/wallet-cordova/src/ios/WalletPlugin.m:249]: Well, this _should_ work so no objections here. If this part if covered by tests, I can check if it works later today.

Comment (enzoc4) [reply bindings/wallet-cordova/src/ios/WalletPlugin.m:249]: well, it should be covered, although if it fails the problem can be in either this function or in the function that gives you the spending counters, as the test sets the spending counters, creates a tx, and asserts over the new counters.

**Review by Zeegomo — APPROVED** [review]
Looks good, maybe thinking we could expose an interface with a default value for the lane in case the client is not interested in advanced counters manipulation

File: bindings/wallet-js/src/lib.rs [hunk 3]
```diff
@@ -144,13 +154,15 @@ impl Wallet {
         proposal: &Proposal,
         choice: u8,
         valid_until: &BlockDate,
+        lane: u8,
```
[the remark this answers is not in the document]
Comment (enzoc4) [reply bindings/wallet-js/src/lib.rs:157]: parameter list?

[the remark this answers is not in the document]
Comment (Zeegomo) [reply bindings/wallet-js/src/lib.rs:157]: There is a doc comment about parameters above the function

**Comment by enzoc4** [comment]
> Looks good, maybe thinking we could expose an interface with a default value for the lane in case the client is not interested in advanced counters manipulation

I'm not sure choosing the lane adds that much complexity, maybe providing a constant with the default like `DEFAULT_LANE` could help? 

I guess the alternative is to provide two functions... My only problem is that I'm not actually sure what would `default` be, since something like doing round robin may be reasonable too.

File: bindings/wallet-js/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (enzoc4) [reply bindings/wallet-js/src/lib.rs:157]: Ah, yeah, good point

**Comment by Zeegomo** [comment]
> I guess the alternative is to provide two functions... My only problem is that I'm not actually sure what would default be, since something like doing round robin may be reasonable too.

I was thinking of an optional value but this works as well.
I think the old behavior, which is the most widely used by other blockchains as well is a sensible default, but this is just a suggestion, we don't have to add this if we think it's not needed

**Pull request merged**